src/corehost/common/trace.h

```diff
@@ -31,6 +33,39 @@ namespace trace
 
     // Returns the currently set callback for error writing
     error_writer_fn get_error_writer();
+
+    void trace(const pal::char_t* format, ...);
+    void trace_error(const pal::char_t* format, ...);
+
+    void println(const pal::char_t* format, ...);
+    void println();
+    void flush();
 };
 
+// Tracing is performed with a macro so that arguments to trace::trace() and
+// trace::trace_error() can be evaluated lazily depending on the current
+// verbosity level.
+#define TRACE_LAZY(Verbosity, Args...)                              \
+    do {                                                            \
+        if (Verbosity == trace::verbosity::Error)                   \
+        {                                                           \
+            /* Error conditions are rare: do not pollute the I$ */  \
+            /* real state with tracing calls. */                    \
+            auto cold_trace = [&]() COLD_FUNCTION NO_INLINE         \
```

Comment:
I kind of like this - I assume it helps the compiler by moving the error related code out into a different method...
You may want to explain the `I$` term - I personally have no idea what it means (I can only guess from context).

Reply:
Arbitrary lambda capture is incredibly expensive and the semantics aren't always obvious. If perf is a concern lambdas should be avoided entirely.

Reply:
(`I$` is "instruction cache"; `$` (cash) sounds similar to "cache", and it's often used interchangeably in this context.  You might see things like `D$` for "data cache", or `L1D$` for the L1 data cache.)